lib/Sema/CheckedCSubst.cpp

```diff
@@ -140,6 +140,16 @@ void Sema::CompleteTypeAppFields(RecordDecl *Incomplete) {
     // TODO: are TypeSouceInfo and InstType in sync?
     FieldDecl *NewField = FieldDecl::Create(Field->getASTContext(), Incomplete, SourceLocation(), SourceLocation(),
       Field->getIdentifier(), InstType, Field->getTypeSourceInfo(), Field->getBitWidth(), Field->isMutable(), Field->getInClassInitStyle());
+
+    // Substitute in the bounds-safe interface type (itype).
+    if (auto IType = Field->getInteropTypeExpr()) {
+      auto InstType = SubstituteTypeArgs(IType->getType(), Incomplete->typeArgs());
+      InteropTypeExpr *NewIType = new (Context) InteropTypeExpr(InstType, SourceLocation(), SourceLocation(), IType->getTypeInfoAsWritten());
+      NewField->setInteropTypeExpr(Context, NewIType);
+    }
+    // No substitution is needed for the bounds expression, because no types appear here.
```

Comment:
Actually, the programmer could write code that uses the type variable in a bounds expression.

Reply:
@dtarditi 
Good point. Can types appear in bound expressions only inside `sizeof()`? If so, should we bother with doing the substitution?
e.g.

Suppose we have

```C
struct Foo _For_any(T) {
  _Array_ptr<int> p: count(sizeof(T*)); // non-sensical as a bound, but let's use it just for the sake of having a type in a bounds expression
};

struct Foo<int> f;
```

If (as things stand right now) we don't do the substitution T->int, then the bound on `f` will be `count(sizeof(T*))`, which is erased to `count(size(void*))`. Ultimately, this is the right value, because sizeof(T*) == sizeof(void*) for all T.

So not doing the substitution means we leave a `T` floating around when it shouldn't be, but the result seems to be correct.

The difficulty in doing the substitution is that we'd need to write an expression traversal function for substitution. Right now the substitution function we have operates on types.

Are you ok with leaving the type variable unchanged?

Reply:
I've replaced the comment to be more accurate. Let me know if you want to do the substitution.